aiosmtpd/smtp.py

```diff
@@ -615,78 +668,93 @@ async def _auth_interact(self, server_message) -> _TriStateType:
     #             the SMTP user
     #      - 'identity' is not always username, depending on the auth mecha-
     #        nism. Might be a session key, a one-time user ID, or any kind of
     #        object, actually.
     #      - If the client provides "=" for username during interaction, the
     #        method MUST return b"" (empty bytes)
+    # 2. (NEW SYSTEM) All auth_* methods must return an AuthResult object.
+    #    For explanation on the object's attributes,
+    #    see the AuthResult class definition.
     # 3. Auth credentials checking is performed in the auth_* methods because
     #    more advanced auth mechanism might not return login+password pair
     #    (see #2 above)
 
-    async def auth_PLAIN(self, _, args: List[str]):
-        loginpassword: _TriStateType
+    async def auth_PLAIN(self, _, args: List[str]) -> AuthResult:
+        login_and_password: _TriStateType
         if len(args) == 1:
             # Trailing space is MANDATORY
             # See https://tools.ietf.org/html/rfc4954#page-4
-            loginpassword = await self._auth_interact("334 ")
-            if loginpassword is MISSING:
-                return
+            login_and_password = await self._auth_interact("334 ")
+            if login_and_password is MISSING:
+                return AuthResult(False)
         else:
             blob = args[1].encode()
             if blob == b"=":
-                loginpassword = None
+                login_and_password = None
             else:
                 try:
-                    loginpassword = b64decode(blob, validate=True)
+                    login_and_password = b64decode(blob, validate=True)
                 except Exception:
                     await self.push("501 5.5.2 Can't decode base64")
-                    return
-        if loginpassword is None:
+                    return AuthResult(False)
+        if login_and_password is None:
             login = password = None
         else:
             try:
-                _, login, password = loginpassword.split(b"\x00")
+                _, login, password = login_and_password.split(b"\x00")
             except ValueError:  # not enough args
                 await self.push("501 5.5.2 Can't split auth value")
-                return
-        if self._auth_callback("PLAIN", login, password):
-            if login is None:
-                login = EMPTYBYTES
-            return login
-        else:
-            return MISSING
+                return AuthResult(False)
+        return self._authenticate("PLAIN", _LoginPassword(login, password))
 
-    async def auth_LOGIN(self, _, args: List[str]):
-        login: _TriStateType
-        # 'User Name\x00'
-        login = await self._auth_interact("334 VXNlciBOYW1lAA==")
+    async def auth_LOGIN(self, _, args: List[str]) -> AuthResult:
+        if len(args) > 1:
+            try:
+                login = b64decode(args[1].encode(), validate=True)
```

Comment:
This piece of code handles the -- actually RFC non-compliant -- `AUTH LOGIN <b64_username>` procedure.

Well, actually there is _no_ RFC for `AUTH LOGIN` ... only [a draft RFC](https://www.ietf.org/archive/id/draft-murchison-sasl-login-00.txt) that never gets adopted/promoted into full RFC status. So in this case, we can be more lenient with implementation of `AUTH LOGIN`.

(Also, Python>=3.5 `smtplib.SMTP` is actually buggy; see [**bpo-27820**](https://bugs.python.org/issue27820))